pkg/iris/eventProducer.go

```diff
@@ -0,0 +1,151 @@
+	created, updated, deleted, err := p.createECSTEvents(data, oldData, configId)
+	if err != nil {
+		return err
+	}
+	ecstEvents := append(created, updated...)
+	ecstEvents = append(ecstEvents, deleted...)
+	if len(ecstEvents) == 0 {
+		return nil
+	}
+	scannedEcstObjectsByte, err := storage.Marshal(ecstEvents)
+	if err != nil {
+		return errors.Wrap(err, "Marshall scanned ECST services")
+	}
+
+	return p.irisApi.PostEcstResults(scannedEcstObjectsByte)
+}
+
+func (p *eventProducer) PostStatus(status []byte) error {
+	return p.irisApi.PostStatus(status)
+}
+
+func (p *eventProducer) createECSTEvents(data []models.Data, oldData []models.DiscoveryEvent, configId string) ([]models.DiscoveryEvent, []models.DiscoveryEvent, []models.DiscoveryEvent, error) {
+	deletedEvents := make([]models.DiscoveryEvent, 0)
+	resultMap := p.createItemMap(data, configId)
+	oldResultMap := p.createOldItemMap(oldData)
+
+	createdEvents, updatedEvents, oldResultMap, err := p.FilterForChangedItems(resultMap, oldResultMap, configId)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	// Create DELETED events
+	for _, oldItem := range oldResultMap {
+		deletedEvents = append(deletedEvents, CreateEcstDiscoveryEvent(EventTypeChange, EventActionDeleted, oldItem.Body.State.Data, p.runId, p.workspaceId, configId))
+	}
+	return createdEvents, updatedEvents, deletedEvents, nil
+
+}
+
+func (p *eventProducer) createItemMap(data []models.Data, configId string) map[string]models.Data {
+	resultMap := map[string]models.Data{}
+	for _, item := range data {
+		// Build unique string hash for discoveryItem
+		id := GenerateId(p.workspaceId, configId, item)
+		resultMap[id] = item
+	}
+	return resultMap
+}
+
+func (p *eventProducer) createOldItemMap(data []models.DiscoveryEvent) map[string]models.DiscoveryEvent {
+	resultMap := map[string]models.DiscoveryEvent{}
+	for _, item := range data {
+		// Use id hash from iris to as a key
+		resultMap[item.HeaderProperties.Id] = item
+	}
+	return resultMap
+}
+
+func GenerateHash(toBeHashed interface{}) (string, error) {
+	serialized, err := json.Marshal(toBeHashed)
+	if err != nil {
+		return "", err
+	}
+	hashed := sha256.Sum256(serialized)
+	return hex.EncodeToString(hashed[:]), nil
+
+}
+
+func (p *eventProducer) FilterForChangedItems(newData map[string]models.Data, oldData map[string]models.DiscoveryEvent, configId string) ([]models.DiscoveryEvent, []models.DiscoveryEvent, map[string]models.DiscoveryEvent, error) {
+	updated := make([]models.DiscoveryEvent, 0)
+	created := make([]models.DiscoveryEvent, 0)
+	for id, newItem := range newData {
+		// if the current element from the freshly discovered items is not in the old results, create an CREATED event
+		if oldItem, ok := oldData[id]; !ok {
+			created = append(created, CreateEcstDiscoveryEvent(EventTypeChange, EventActionCreated, newItem, p.runId, p.workspaceId, configId))
+			// if item exists in old and new result sets but their payloads differ
```

Comment:
what exactly does this mean? it isn't very clear for me

Reply:
If the item existed before, compare if the data of the two items are identical or not

Reply:
Sorry for being nitpicky but I think it is more straightforward to say "Compare payload if the item exists in both old and new data" 😅

Reply:
yes already updated it to something like that 👍

Reply:
Please check again